### ISSUE TITLE
sql: remove or reimplement `experimental_unique_int`

### ISSUE THREAD

**Issue by bdarnell** [cockroachdb/cockroach#4183]
`experimental_unique_int` is a trap: it breaks down when node IDs pass 2^15. It is only safe to use when either A) you know you will not have that many nodes (but multi-tenant hosting makes this impossible to know) or B) your ID generation is infrequent enough that collisions are unlikely even with an imprecise generator (although in that case you can probably get away with a plain timestamp instead of a function that mixes in the node ID).

`experimental_unique_bytes` is nearly always preferable (and only costs a few more bytes), but experience has shown that when both functions exist, developers will usually choose the `int` version. I think we should get rid of it to encourage the use of appropriately-sized `bytes` fields whenever uniqueness is required.

Alternately, we could change the implementation to provide somewhat better uniqueness properties within the size of an integer. Simply truncating the node ID at 15 bits instead of allowing it to spill into the timestamp would help. Or we could make the ID random to enhance uniqueness (which is in the name of the function) at the expense of k-sortedness (which is not).

**Comment by JackKrupansky** [comment]
What are the intended use cases for experimental_unique_int? I mean, how are people expected to use it?

Currently, I see only two uses in the current code base...

It used for the rowid for CREATE TABLE when a primary key is not present. In that particular code  it only needs to be unique within the table, which is shared across all nodes anyway.

It is also used by the rangelog table, but there the PK  requires the timestamp and range, so once again it is not clear that the node is required for uniqueness outside of the context in which it seems to be used. Or is there code somewhere that decodes the node from the uniqueID? Actually, I currently don't find any references to it at all.

**Comment by bdarnell** [comment]
The intention is to provide functionality equivalent to mysql's AUTO_INCREMENT or postgresql's SERIAL columns. However, while a single-node database can provide several useful properties (uniqueness and monotonicity) cheaply (and fitting within a 64-bit int), attempting to do the same in a distributed database requires compromise. `experimental_unique_int` is only k-sorted instead of monotonic and has some risk of collisions. `experimental_unique_bytes` has stronger uniqueness guarantees, but it can't fit in an int64 (and people really like their IDs to be numbers instead of blobs for no good reason).

In the rangelog table it's just being used as a tiebreaker in case there is a collision with timestamp and range ID. This has very weak requirements for uniqueness so it's fine with the current implementation (in fact it would be fine with a much weaker implementation like a small random number). The rowid case is the more concerning one; a table without a primary key could get large enough that collisions are a problem, so this is where it's important to provide stronger guarantees of uniqueness.

**Comment by JackKrupansky** [comment]
-- Jack Krupansky

**Comment by JackKrupansky** [comment]
Answering my own question - the node ID is "needed" for rowid to avoid the
need to synchronize when inserts are made on different nodes - the main
issue standing in the way of supporting SERIAL and SEQUENCE.

-- Jack Krupansky

**Comment by petermattis** [comment]
I think we need to provide a way to generate unique 64-bit ints as sometimes there are external constraints the prevent using a larger type. For example, the [filesystem example](https://github.com/cockroachdb/examples-go/blob/master/filesystem/fs.go#L278) uses `experimental_unique_int` to allocate inode IDs which need to fit in a 64-bit integer. 

This isn't to say that we need to the current semantics or that we should live with the current limitations. If we relaxed the desire for a k-sorted integer, we could give more bits to the node-ID.

**Comment by petermattis** [comment]
As I mentioned in my previous comment, I think we need to support unique 64-bit int generation as external requirements sometimes force usage of an int vs bytes. I see two ways forward here: try to adapt the current timestamp-based unique-int generation to something more palatable, or switch to a random-based generator.

The current time-based generator uses 48-bits for the timestamp at 10-microsecond precision and 15-bits for the node-id. I think we can increase the bits for the node-id to 18 (262k nodes) and reduce the timestamp precision to 100-microseconds. The generator also avoids negative numbers. If we allow negative numbers we could give another bit to the node-id (524k) which seems more than enough for now. When we start reaching clusters larger than that we can have a crack-team work on a v2 unique-int generator.

A random-based generator has its own difficulties which I'm not sure how to avoid. We can't just use a random 64-bit int as you'd expect collisions after 4 billion have been generated. So perhaps we combine the node-id with a random portion. How do we generate the random portion? Pseudo-random permutations of integers starting with a particular seed are possible, but they seem to require storing the state of the permutation frequently which is awkward. Is there some other alternative I'm not seeing here?

**Comment by JackKrupansky** [comment]
Is the issue the total cluster size (active nodes) or the fact that over time as nodes die and/o upgraded and replaced that the node IDs increase way beyond the number of actually active nodes? (Might there eventually be any node ID reuse, intentional recycling, over time?)

From a practical perspective, I have a hard time feeling like you need to worry about clusters with more than a few hundred nodes in the next few years. Last I heard (September) Apple had 125,000 Cassandra nodes, but on a number of disjoint clusters (different applications with different data.)

I was thinking that a slightly coarser time stamp would be tolerable, but then the immediate question is what the maximal rate of inserts could be on a given node, where you have multiple requests in parallel. But then I realized that if you have multiple requests appending to the same table you probably hit some transaction concurrency so that the transactions need to get restarted anyway and the 250 ms time interval kicks in so that 100 usec is more than enough. In short, how many inserts per second do you want to design for on a single node and with what transaction timing constraint?

Oops... and then there are parallel stores on the same node, which presumably can be inserting at exactly the same time on parallel requests. Maybe it depends on what experimental_unique_int is being used for - a unique rowid for a single table or just some random field that isn't necessarioy a table PK. Parallel inserts in different tables have no conflict per se - as long as the contract for experimental_unique_int presumes that multiple concurrent invocations are for independent usages rather than for the same table. The contract should be made explicit.

In short, I think both the current and proposed form of 64-bit unique int is fine, although I personally still lean towards a per-table unique counter. I'm not excited about random - maybe it is great for some use cases, but it would probably be better for the user to specify their requirements as to what ordering they need or don't care about (maybe a SERIAL vs. DISPERSED keyword, and even SERIAL PRECISE vs. SERIAL MONOTONIC, which there later indicates that i just needs to increase and be ordered but not a precise counter.)

**Comment by petermattis** [comment]
@bdarnell You filed the original issue. Can you weigh in here?

**Comment by bdarnell** [comment]
My concern with the current implementation is that it has a cliff: once you reach 32k node IDs, the collision rate goes up dramatically, especially considering that HLC updates can cause timestamps to cluster across nodes. With purely random IDs (I wasn't thinking about mixing node IDs with randomness; I don't think that really helps), you do expect a collision around 4B IDs, but you don't expect a second until around 8B, etc. I don't think we should use random IDs, though - 4B entries without collision just isn't enough for a distributed database. 

This discussion is partly about a simple bug: If we xor'd the node id and shifted timestamp together instead of or'ing them, there wouldn't be a cliff and I think we'd have an ID generator that is about as good as we can get given the 64-bit constraint. I think that's acceptable, if we conclude that we do need to provide such a function.

But the bigger issue is that people really like 64-bit IDs because they're used to them from non-distributed databases where they can be used reliably. In our case small IDs have caveats that aren't present in non-distributed DBs, so we should think about how we can discourage people from using 64-bit IDs except when they have a true need imposed by an external system (like the filesystem example). And when they do have such a need, are they better off paying the cost of a centralized counter or can they accept the (small) risk of collision? We've seen from our own experience that we use `experimental_unique_int` when `experimental_unique_bytes` would have been the better choice. Maybe this is all moot since we've already decided to use int64 for our own internal row IDs.

**Comment by petermattis** [comment]
I think if we don't provide a unique 64-bit ID generator users will end up creating one themselves, poorly. 

How about this plan: I switch to xor'ing the node-id instead of or'ing it and I allow the use of the full 64-bit value and dedicate the extra bit to the node-id allowing up to 64k before we hit trouble. 

Another thought for how we can allow larger node-ids is to say that any node with an ID larger than 64K would talk to a neighbor with a node ID less than 64K. Pick a close neighbor and the RPC overhead will be small. We could also restrict node IDs to be less than 64K until we figure out a longer term solution here. Then the user wouldn't be surprised by a collision but would see an error when the node was being added.

**Comment by bdarnell** [comment]
SGTM. I'm a little concerned about using the high bit since I could imagine a lot of systems having trouble with negative numbers (and we don't have unsigned ints).

I'm not sure that talking to a close neighbor with a low ID will be possible when large clusters arrive. You'd end up with a new datacenter consisting solely of high IDs. It would be better to designate certain nodes (or ranges) as the ID generators independently of their node IDs. But we can figure that out later.

**Comment by JackKrupansky** [comment]
For large clusters, it would be great to have the ability to allocate and reserve a moderate-size block of unique IDs for each node, like maybe 1M at a time, and then request a new block as needed. The occasional overhead of synchronizing to allocate a new block of IDs would seem insignificant.

Now you are effectively allocating 48-bits of ID permanently for each node when in reality you would rarely need more than 32 bits for an incremented ID for each node.

**Comment by petermattis** [comment]
I don't think there is anything left to do here. `experimental_unique_int` is now `unique_rowid`. We're not using the high-bit and we're xor-ing (not or-ing) in the nodeID bits.
